lib/Target/CBackend/CBackend.cpp

```diff
@@ -3158,14 +3148,54 @@ void CWriter::printModuleTypes(raw_ostream &Out) {
 
   Out << "\n/* Function definitions */\n";
 
+  struct FunctionDefinition {
+    FunctionType *FT;
+    std::vector<FunctionType *> Dependencies;
+    std::string NameToPrint;
+  };
+
+  std::vector<FunctionDefinition> FunctionTypeDefinitions;
+  // Copy Function Types into indexable container
   for (auto &I : UnnamedFunctionIDs) {
-    Out << '\n';
-    std::pair<FunctionType *, std::pair<AttributeList, CallingConv::ID>> F =
-        I.first;
-    if (F.second.first.isEmpty() && F.second.second == CallingConv::C)
-      if (!TypesPrinted.insert(F.first).second)
-        continue; // already printed this above
-    printFunctionDeclaration(Out, F.first, F.second);
+    const auto &F = I.first;
+    FunctionType *FT = F.first;
+    std::vector<FunctionType *> FDeps;
+    for (const auto P : F.first->params()) {
+      if (P->isPointerTy()) {
+        PointerType *PP = dyn_cast<PointerType>(P);
+        if (PP->getElementType()->isFunctionTy()) {
+          FDeps.push_back(dyn_cast<FunctionType>(PP->getElementType()));
+        }
+      }
+    }
+    std::string DeclString;
+    raw_string_ostream TmpOut(DeclString);
+    printFunctionDeclaration(TmpOut, F.first, F.second);
+    TmpOut.flush();
+    FunctionTypeDefinitions.emplace_back(
+        FunctionDefinition{FT, FDeps, DeclString});
+  }
+
+  // Sort function types
+  TopologicalSorter Sorter(FunctionTypeDefinitions.size());
+  DenseMap<FunctionType *, int> TopologicalSortMap;
+  // Add Vertices
+  for (unsigned I = 0; I < FunctionTypeDefinitions.size(); I++) {
+    TopologicalSortMap[FunctionTypeDefinitions[I].FT] = Sorter.addVertex(I);
+  }
+  // Add Edges
+  for (unsigned I = 0; I < FunctionTypeDefinitions.size(); I++) {
+    const auto &Dependencies = FunctionTypeDefinitions[I].Dependencies;
+    for (unsigned J = 0; J < Dependencies.size(); J++) {
+      Sorter.addEdge(I, TopologicalSortMap[Dependencies[J]]);
+    }
+  }
+  Optional<std::vector<int>> TopologicalSortResult = Sorter.sort();
+  if (!TopologicalSortResult.hasValue()) {
+    errorWithMessage("Cyclic dependencies in function definitions");
```

Comment:
I thought this seemed likely to be reachable, but I couldn't think of how to get LLVM to emit something self-referential. The closest I got was the following, but LLVM just emits an empty struct and a bitcast:
```C
$ clang -emit-llvm -S -o - -x c - -O1
struct selfcall;
typedef void (fselfcall)(struct selfcall f);
struct selfcall {
    fselfcall *f;
};
void infiniteloop(struct selfcall f) {
    return f.f(f);
}
```
```llvm
; ModuleID = '-'
source_filename = "-"
target datalayout = "e-m:o-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-apple-macosx10.13.0"

; Function Attrs: nounwind ssp uwtable
define void @infiniteloop({}*) local_unnamed_addr #0 {
  %2 = bitcast {}* %0 to void ({}*)*
  tail call void %2({}* %0) #1
  ret void
}

attributes #0 = { nounwind ssp uwtable "correctly-rounded-divide-sqrt-fp-math"="false" "disable-tail-calls"="false" "less-precise-fpmad"="false" "no-frame-pointer-elim"="true" "no-frame-pointer-elim-non-leaf" "no-infs-fp-math"="false" "no-jump-tables"="false" "no-nans-fp-math"="false" "no-signed-zeros-fp-math"="false" "no-trapping-math"="false" "stack-protector-buffer-size"="8" "target-cpu"="penryn" "target-features"="+cx16,+fxsr,+mmx,+sahf,+sse,+sse2,+sse3,+sse4.1,+ssse3,+x87" "unsafe-fp-math"="false" "use-soft-float"="false" }
attributes #1 = { nounwind }

!llvm.module.flags = !{!0, !1}
!llvm.ident = !{!2}

!0 = !{i32 1, !"wchar_size", i32 4}
!1 = !{i32 7, !"PIC Level", i32 2}
!2 = !{!"Apple LLVM version 10.0.0 (clang-1000.11.45.5)"}
```

Reply:
I do not think it is possible to create cycles with functions taking function pointers as argument in llvm IR. 
[A blog post about the llvm type system says : ](http://blog.llvm.org/2011/11/llvm-30-type-system-rewrite.html) 
> Only struct types can be recursive. 
In the previous type system, an OpaqueType could be resolved to any arbitrary type, allowing such oddities as "%t1 = type %t1*", which is a pointer to itself. In the new type system, only IR structure types can have their body missing, so it is impossible to create a recursive type that doesn't involve a struct. 

The only way to create something recursive seems to be a an opaque struct, like :
```LLVM
%TOpaque = type opaque
%TReal = type { i32 (%TOpaque*) *} 

; Function Attrs: nounwind uwtable
define i32 @call4(%TOpaque*) unnamed_addr #0 {
  %bcast = bitcast %TOpaque* %0 to %TReal*
  %2 = getelementptr %TReal, %TReal* %bcast, i32 0, i32 0
  %3 = load i32 (%TOpaque*) *, i32 (%TOpaque*) ** %2
  %4 = call i32 %3(%TOpaque* %0) #2
  ret i32 %4
}
```

But this should be safe, because the opaque struct argument gets converted to a void pointer.

Sadly llvm-cbe fails in this case, because it generates (But this is easy to fix):
```C
/* Types Declarations */
struct l_struct_TReal;

/* Types Definitions */
struct l_struct_TReal {
  l_fptr_1* field0;
};

/* Function definitions */
typedef uint32_t l_fptr_1(void*);
```

Reply:
This code now also works

Reply:
The recursive code above does now work too.
@vtjnash It would be nice if you could look at this and my other merge request